phase_1_20160807/hugo_20160804_lifecycle_launch_mode.md

```diff
@@ -0,0 +1,396 @@
+Activity / Fragment 的生命周期是每个 Android 开发者最最基础的知识点。所以特别有必要自己整理一番。总看别人博客和书上的死知识，还不如自己动手实践，然后输出要印象深刻，理解透彻。
+
+<!-- more -->
+
+<!-- toc -->
+
+- [Activity 生命周期](#activity-生命周期)
+  - [正常情况下的生命周期分析](#正常情况下的生命周期分析)
+  - [异常状态下的生命周期](#异常状态下的生命周期)
+- [Fragment](#fragment)
+  - [普通的 Fragment](#普通的-fragment)
+  - [ViewPager 中的 Fragment](#viewpager-中的-fragment)
+- [启动模式](#启动模式)
+  - [Activity 的四种启动模式](#activity-的四种启动模式)
+  - [具体实践](#具体实践)
+- [参考文档](#参考文档)
+
+<!-- tocstop -->
+
+### Activity 生命周期
+
+
+#### 正常情况下的生命周期分析
+
+![22:39:52.jpg](http://ww2.sinaimg.cn/large/006tKfTcgw1f6i3dzl70yj30fj0j2q41.jpg)
+
+1. 针对一个特定的 Activity ，第一次启动，回调如下：`onCreate` —-> `onStart` —-> `onResume`
+>Log 日志
+>D/KLog: (MainActivity.java:19) onCreate
+>D/KLog: (MainActivity.java:44) onStart
+>D/KLog: (MainActivity.java:62) onResume
+
+1. 切换回到桌面的时候，回调如下：`onPause` —-> `onStop`
+> Log 日志
+> D/KLog: (MainActivity.java:50) onPause
+> D/KLog: (MainActivity.java:68) onStop
+
+1. Back 键退出的话，最后会 `onDestroy`
+
+2. 启动一个新的 Activity , 我们看看两个 Activity 的生命周期：
+> Log 日志
+> D/KLog: (MainActivity.java:50) onPause
+> D/KLog: (OtherActivity.java:25) onCreate
+> D/KLog: (OtherActivity.java:31) onStart
+> D/KLog: (OtherActivity.java:49) onResume
+> D/KLog: (MainActivity.java:68) onStop
+> 可以得到顺序是：`onPause(A)` —-> `onCreate(B)` —-> `onStart(B)` —->  `onResume(B)` —-> `onStop(A)`
+
+1. 这个时候我们 Back 回到第一个 Activity 时发生的回调：
+> Log 日志
+> D/KLog: (OtherActivity.java:37) onPause
+> D/KLog: (MainActivity.java:56) onRestart
+> D/KLog: (MainActivity.java:44) onStart
+> D/KLog: (MainActivity.java:62) onResume
+> D/KLog: (OtherActivity.java:55) onStop
+> D/KLog: (OtherActivity.java:61) onDestroy
+> 可以得到顺序是： `onPause(B)` —-> `onRestart(A)` —-> `onStart(A)` —-> `onResume(A)` —-> `onStop(B)` —-> `onDestroy(B)`
+
+1. 如果我在 B Activity 中的 `onCreate` 回调中直接 `finish()`：
+> Log 日志
+> D/KLog: (MainActivity.java:50) onPause
+> D/KLog: (OtherActivity.java:25) onCreate
+> D/KLog: (MainActivity.java:62) onResume
+> D/KLog: (OtherActivity.java:62) onDestroy
+> 我们发现 B Activity 只会执行 `onCreate` 和 `onDestroy`。
+
+1. 接下来我们启动一个特殊的 Activity （半透明或者对话框样式）到关闭它：
+> Log 日志
+> D/MainActivity: onPause
+> D/DialogActivity: onCreate
+> D/DialogActivity: onStart
+> D/DialogActivity: onResume
+> D/DialogActivity: onPause
+> D/MainActivity: onResume
+> D/DialogActivity: onStop
+> D/DialogActivity: onDestroy
+
```

Comment:
嗯嗯 好


> 在 2016年8月5日，下午10:46，allenwu <notifications@github.com> 写道：
> 
> In phase_1_20160807/hugo_20160804_lifecycle_launch_mode.md <https://github.com/itsCoder/weeklyblog/pull/18#discussion_r73704109>:
> 
> > +> D/KLog: (OtherActivity.java:25) onCreate
> > +> D/KLog: (MainActivity.java:62) onResume
> > +> D/KLog: (OtherActivity.java:62) onDestroy
> > +> 我们发现 B Activity 只会执行 `onCreate` 和 `onDestroy`。
> > +
> > +1. 接下来我们启动一个特殊的 Activity （半透明或者对话框样式）到关闭它：
> > +> Log 日志
> > +> D/MainActivity: onPause
> > +> D/DialogActivity: onCreate
> > +> D/DialogActivity: onStart
> > +> D/DialogActivity: onResume
> > +> D/DialogActivity: onPause
> > +> D/MainActivity: onResume
> > +> D/DialogActivity: onStop
> > +> D/DialogActivity: onDestroy
> > +
> 可以用 代码块把日志包起来。试试看，应该会换行。
> 
> ''' java
> 
> D/MainActivity: onPause
> D/DialogActivity: onCreate
> D/DialogActivity: onStart
> D/DialogActivity: onResume
> D/DialogActivity: onPause
> D/MainActivity: onResume
> D/DialogActivity: onStop
> D/DialogActivity: onDestroy
> '''
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/itsCoder/weeklyblog/pull/18/files/28b03b3d6e3476906bbeeb1a9059db052e49a163#r73704109>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALIEyjgj0T3ZD_YiF0iJzOa8h-3MD3joks5qc0zBgaJpZM4Jcpjd>.
>